pkg/buildqueue/buildqueue.go

```diff
@@ -0,0 +1,580 @@
+	exclusiveBuilds buildGroup
+
+	// waitingBuilders is used to wake up builders waiting for their
+	// condition to be met
+	waitingBuilders *sync.Cond
+}
+
+// NewBuildQueue returns a new build queue
+func NewBuildQueue(name string) *BuildQueue {
+	q := &BuildQueue{
+		name:             name,
+		buildStatus:      buildStatusMap{},
+		workerBuildQueue: make(chan Builder, buildQueueSize),
+		stopWorker:       make(chan struct{}, 0),
+		regularBuilds: buildGroup{
+			condition: func(q *BuildQueue) bool {
+				return q.exclusiveBuilds.waiting == 0 && q.exclusiveBuilds.running == 0
+			},
+		},
+		exclusiveBuilds: buildGroup{
+			condition: func(q *BuildQueue) bool {
+				return q.exclusiveBuilds.running == 0 && q.regularBuilds.running == 0
+			},
+		},
+	}
+
+	q.waitingBuilders = sync.NewCond(&q.mutex)
+
+	nWorkers := numWorkerThreads()
+	for w := 0; w < nWorkers; w++ {
+		go q.runWorker()
+	}
+
+	return q
+}
+
+// Stop stops the build queue and terminates all workers
+func (q *BuildQueue) Stop() {
+	close(q.stopWorker)
+
+	waitForCleanup := sync.WaitGroup{}
+
+	q.mutex.Lock()
+	for _, status := range q.buildStatus {
+		waitForCleanup.Add(1)
+		go func(b Builder) {
+			q.Drain(b)
+			waitForCleanup.Done()
+		}(status.builder)
+	}
+	q.mutex.Unlock()
+
+	waitForCleanup.Wait()
+}
+
+// numWorkerThreads returns the number of worker threads to use. This does not
+// apply to exclusive builds.
+func numWorkerThreads() int {
+	ncpu := runtime.NumCPU()
+
+	if ncpu < minWorkerThreads {
+		return minWorkerThreads
+	}
+	return ncpu
+}
+
+// Buildable is an object that is buildable
+type Buildable interface {
+	// GetUUID must return a unique UUID of the object
+	GetUUID() string
+}
+
+// Builder is an object that can build itself. A builder must also be Buildable
+type Builder interface {
+	Buildable
+
+	// BuildQueued is called every time a builder is enqueued
+	BuildQueued()
+
+	// BuildsDequeued is called when a scheduled builder has been dequeued
+	// for building or has been cancelled. The nbuilds accounts for the
+	// number of builders folded together while the builder was queued.
+	BuildsDequeued(nbuilds int, cancelled bool)
+
+	// Build must build the Buildable. When a builder was cancelled,
+	// Build() will not be called. When a builder is folded into an
+	// existing builder, Build() is only called on the first builder.
+	Build() error
+}
+
+// BuildNotification is a channel to receive the status of a build. The channel
+// will receive true if the build was successful. The channel is buffered and
+// is guaranteed to only be written to exactly once and is then immediately
+// closed. It is up to the caller of Enqueue() to decide whether to read on the
+// channel or not.
+type BuildNotification <-chan bool
+
+func newBuildNotification() chan bool {
+	return make(chan bool, 1)
+}
+
+// build is what is returned by Enqueue() to provide information about the
+// build
+type build struct {
+	builder Builder
+
+	// uuid is the UUID of the Buildable
```

Comment:
Buildable --> Builder?

Reply:
The UUID is derived from the Buildable interface